src/data/data.cc

```diff
@@ -829,48 +830,68 @@ uint64_t SparsePage::Push(const AdapterBatchT& batch, float missing, int nthread
           last_line.GetElement(last_line.Size() - 1).row_idx - base_rowid;
     }
   }
-  builder.InitBudget(expected_rows, nthread);
+  size_t batch_size = batch.Size();
+  const size_t thread_size = batch_size/nthread;
```

Comment:
Do you need to roundup?

Reply:
with current `begin`, `end` calculation there is no need to roundup